conans/test/integration/conan_env_test.py

```diff
@@ -9,10 +9,117 @@
+    exports_sources = "*"
+
+    def build(self):
+        cmake = CMake(self)
+        self.run('cmake %s' % cmake.command_line)
+        self.run("cmake --build . %s" % cmake.build_config)
+
+    def package(self):
+        self.copy("*.h", dst="include", src="hello")
+        self.copy("*.lib", dst="lib", keep_path=False)
+        self.copy("*.dll", dst="bin", keep_path=False)
+        self.copy("*.so", dst="lib", keep_path=False)
+        self.copy("*.dylib", dst="lib", keep_path=False)
+        self.copy("*main*", dst="bin", keep_path=False)
+
+"""
+        cmakelists =  """
+project(mytest)
+set(CMAKE_CXX_COMPILER_WORKS 1)
+set(CMAKE_CXX_ABI_COMPILED 1)
+
+SET(CMAKE_SKIP_RPATH 1)
+ADD_LIBRARY(hello SHARED hello.c)
+ADD_EXECUTABLE(main main.c)
+TARGET_LINK_LIBRARIES(main hello)
+"""
+
+        hello_h = """#pragma once
+#ifdef WIN32
+  #define HELLO_EXPORT __declspec(dllexport)
+#else
+  #define HELLO_EXPORT
+#endif
+
+HELLO_EXPORT void hello();
+"""
+
+        client = TestClient()
+        files = {CONANFILE: conanfile,
+                 "CMakeLists.txt": cmakelists,
+                 "hello.c": '#include "hello.h"\nvoid hello(){\nreturn;}',
+                 "hello.h": hello_h,
+                 "main.c": '#include "hello.h"\nint main(){\nhello();\nreturn 0;\n}'}
+
+        client.save(files)
+        client.run("export conan/stable")
+        client.run("install lib/1.0@conan/stable -o lib:shared=True --build missing")
+        client.save({"conanfile.txt": '''
+[requires]
+lib/1.0@conan/stable
+[generators]
+virtualrunenv
+[imports]
+bin, * -> ./bin
+lib, * -> ./bin
+'''}, clean_first=True)
+
+        client.run("install .")
+        # Break possible rpaths built in the exe with absolute paths
+        os.rename(os.path.join(client.current_folder, "bin"),
+                  os.path.join(client.current_folder, "bin2"))
+
+        with tools.chdir(os.path.join(client.current_folder, "bin2")):
+            if platform.system() == "Windows":
+                self.assertEqual(os.system("main.exe"), 0)
+            elif platform.system() == "Darwin":
+                self.assertEqual(os.system("./main"), 0)
+            else:
+                self.assertNotEqual(os.system("./main"), 0)
+
+            # If we move the shared library it won't work, at least we use the virtualrunenv
+            os.mkdir(os.path.join(client.current_folder, "bin2", "subdir"))
+            name = {"Darwin": "libhello.dylib",
+                    "Windows": "hello.dll"}.get(platform.system(), "libhello.so")
+
+            os.rename(os.path.join(client.current_folder, "bin2", name),
+                      os.path.join(client.current_folder, "bin2", "subdir", name))
+
+            if platform.system() == "Windows":
+                self.assertNotEqual(os.system("main.exe"), 0)
+            elif platform.system() == "Darwin":
+                self.assertNotEqual(os.system("./main"), 0)
+            else:
+                self.assertNotEqual(os.system("./main"), 0)
```

Comment:
Not sure why the others have changed (failed) and Linux doesn't change

Reply:
Now the shared library is not even in the current directory, so it can't work anyway. I tested it to be sure that no rpath is messing.